src/System.Linq/src/System/Linq/Enumerable.cs

```diff
@@ -2015,31 +2119,68 @@ public static decimal Min<TSource>(this IEnumerable<TSource> source, Func<TSourc
         public static int Max(this IEnumerable<int> source)
         {
             if (source == null) throw Error.ArgumentNull("source");
-            int value = 0;
-            bool hasValue = false;
-            foreach (int x in source)
+            int value;
+            using (IEnumerator<int> e = source.GetEnumerator())
             {
-                if (hasValue)
+                if (!e.MoveNext()) throw Error.NoElements();
+                value = e.Current;
+                while (e.MoveNext())
                 {
+                    int x = e.Current;
                     if (x > value) value = x;
                 }
-                else
-                {
-                    value = x;
-                    hasValue = true;
-                }
             }
-            if (hasValue) return value;
-            throw Error.NoElements();
+            return value;
         }
 
         public static int? Max(this IEnumerable<int?> source)
         {
             if (source == null) throw Error.ArgumentNull("source");
-            int? value = null;
-            foreach (int? x in source)
+            int? value;
+            using (IEnumerator<int?> e = source.GetEnumerator())
             {
-                if (value == null || x > value) value = x;
+                if (!e.MoveNext()) return null;
+                value = e.Current;
+                while (!value.HasValue)
+                {
+                    if (!e.MoveNext()) return value;
+                    value = e.Current;
+                }
+                int valueVal = value.GetValueOrDefault();
+                if (valueVal >= 0)
+                {
+                    // We can fast-path this case where we know HasValue will
+                    // never affect the outcome, without constantly checking
+                    // if we're in such a state. A similar fast-path could
+                    // be done for Min on negative values, but as all-positive
+                    // or mostly-positive values are quite common in real-world
+                    // uses, it's only been done in this direction.
```

Comment:
I decided though that doing it everywhere it could be applied would be too clever for its own good, and restricted it to non-negative integers on `Max`.